### ISSUE TITLE
음악 분류

### ISSUE THREAD

**Issue by rkdogo08** [rkdogo08/Quiz#2]
## 학습 데이터 준비
* 함수`labeling()`은 문자열 배열로 저장된 데이터의 라벨`label`을 입력으로 받아 정수로 저장된 라벨`Y`을 반환하는 함수이다. `label` 중에 `arr[i] `와 같은 인덱스만 `True`로 반환하여 `mask`에 저장하고 `Y[mask]`에 라벨 번호` i `를 저장하여 정수 라벨 배열을 만든다.

```
def labeling(label):
    arr=['blues','classical','country','disco','hiphop','jazz','metal','pop','reggae','rock']
    Y=np.zeros_like(label)
    for i in range(10):
        mask=(label==arr[i])
        Y[mask]=i

    return Y
```
* csv파일로 저장된 데이터를 pandas를 이용하여 읽는다. 데이터는 30개의 행으로 구성되어 있으며 파일 이름과 라벨을 제외한 28개의 피쳐는 `scale()`함수를 이용하여 피쳐 별로 각각 정규화하여 저장한다. 
`labeling() `함수를 호출하여 라벨을 만들어준다.
```
df_data = pd.read_csv('/kaggle/input/2019-pr-midterm-musicclassification/data_train.csv')

feature=['tempo','beats','chroma_stft','rmse','spectral_centroid','spectral_bandwidth','rolloff','zero_crossing_rate','mfcc1','mfcc2','mfcc3','mfcc4','mfcc5','mfcc6','mfcc7','mfcc8','mfcc9','mfcc10','mfcc11','mfcc12','mfcc13','mfcc14','mfcc15','mfcc16','mfcc17','mfcc18','mfcc19','mfcc20']

for i in feature:
    df_data[i] = scale(df_data[i])

sample=950
x=[]
for i in range(sample):
    x.append(np.array(df_data.ix[i][1:29]))
X=np.array(x)

label=np.array(df_data['label'][0:sample])
Y=labeling(label)
Y=Y.astype('int') 
```

* Train set과 validation set은 0.75대 0.25로 나누어 학습을 진행하였다.

```
X_train,X_val,y_train,y_val=train_test_split(X,Y,test_size=0.25,random_state=42)

```

## KNeighborsClassifier과 GridSearchCV를 이용한 모델 설계
* KNeighborsClassifier를 이용하여 모델을 설계하였다. `n_neighbors`는 결과 예측시 몇개의 이웃 데이터를 가지고 투표할 것인가에 대한 파라미터이다. GridSearchCV를 이용하여 가장 정확도가 높은 `n_neighbors` 파라미터를 구하고 `best_estimator_`를 통해 모델을 저장한다.
```
knn = KNeighborsClassifier()
param_grid = {'n_neighbors': np.arange(1, 10)}
grid = GridSearchCV(knn, param_grid, cv=5)
grid.fit(X_train,y_train)
model = grid.best_estimator_
y_fit = model.predict(X_val)
```

## 테스트 데이터 가공
* 학습데이터와 같이 데이터들을 각 피쳐별로 정규화 해준다. 별도의 라벨링 작업은 필요 없다.
```
df_data_T = pd.read_csv('/kaggle/input/2019-pr-midterm-musicclassification/data_test.csv')

for i in feature:
    df_data_T[i] = scale(df_data_T[i])

sample_T=50
X_test=np.zeros(28)
for i in range(sample_T):
    X_test=np.vstack((X_test,np.array(df_data_T.ix[i][1:29])))
X_test=np.delete(X_test,0,axis=0)

```

## 학습한 모델을 이용하여 결과 예측
* Train set을 이용해 학습한 모델을 통해 테스트 데이터의 결과를 예측한다.

```
result=model.predict(X_test)
```

## 결과 csv파일 저장
* 캐글 제출 형식을 맞추기 위해 0번째 행의 index 이름을 "id"로 바꾸고 데이터에 'label'`를 넣어주고 data frame을 csv파일로 변환하여 저장한다.

```
import pandas as pd

result=np.append(['label'],result)
df = pd.DataFrame(result,columns=[' '])
df=df.rename(index={0:"id"})
df.to_csv('results-yk-v2.csv',index=True, header=False)
df
```
